From 643e6799943ab3e3a203f7f41b20f90e63b3ba7a Mon Sep 17 00:00:00 2001
From: Nobuyoshi Nakada <nobu@ruby-lang.org>
Date: Wed, 21 Oct 2026 20:19:58 +0900
Subject: [PATCH] [Bug #22458] Fix out-of-bound in `Time#strftime`

(cherry picked from commit 5c52c2a37bed5507c1898486cdf7eba736296ed4)

Backport to 3.3: upstream also reworked the capacity-doubling loop in
resize_buffer(), which 3.3 does not have (its resize_buffer() computes
len + n * 2 once and checks that directly), so that hunk is dropped and
3.3's resize_buffer() is kept as is. The %z/%:z/%::z/%:::z NEEDS hunks,
which fix the out-of-bound write, are unchanged from upstream.
---
 strftime.c             | 27 ++++++++++++++-------------
 test/ruby/test_time.rb |  7 +++++++
 2 files changed, 21 insertions(+), 13 deletions(-)

diff --git a/strftime.c b/strftime.c
index 33e7d3fdb8..121eb1af96 100644
--- a/strftime.c
+++ b/strftime.c
@@ -555,34 +555,35 @@ rb_strftime_with_timespec(VALUE ftime, const char *format, size_t format_len,
 			else {
 				sign = +1;
 			}
-                        switch (colons) {
+#define NEEDS_FOR_ZONE(fmt) do { \
+				const int zone_size = rb_strlen_lit(fmt); \
+				if (precision < zone_size) precision = zone_size; \
+				NEEDS(precision); \
+				precision -= zone_size - rb_strlen_lit("hh"); \
+			} while (0)
+
+                       switch (colons) {
 			case 0: /* %z -> +hhmm */
-				precision = precision <= 5 ? 2 : precision-3;
-				NEEDS(precision + 3);
+				NEEDS_FOR_ZONE("+hhmm");
 				break;
 
 			case 1: /* %:z -> +hh:mm */
-				precision = precision <= 6 ? 2 : precision-4;
-				NEEDS(precision + 4);
+				NEEDS_FOR_ZONE("+hh:mm");
 				break;
 
 			case 2: /* %::z -> +hh:mm:ss */
-				precision = precision <= 9 ? 2 : precision-7;
-				NEEDS(precision + 7);
+				NEEDS_FOR_ZONE("+hh:mm:ss");
 				break;
 
 			case 3: /* %:::z -> +hh[:mm[:ss]] */
 				if (off % 3600 == 0) {
-					precision = precision <= 3 ? 2 : precision-1;
-					NEEDS(precision + 3);
+					NEEDS_FOR_ZONE("+hh");
 				}
 				else if (off % 60 == 0) {
-					precision = precision <= 6 ? 2 : precision-4;
-					NEEDS(precision + 4);
+					NEEDS_FOR_ZONE("+hh:mm");
 				}
 				else {
-					precision = precision <= 9 ? 2 : precision-7;
-					NEEDS(precision + 9);
+					NEEDS_FOR_ZONE("+hh:mm:ss");
 				}
 				break;
 
diff --git a/test/ruby/test_time.rb b/test/ruby/test_time.rb
index 2a541bbe8c..9934148449 100644
--- a/test/ruby/test_time.rb
+++ b/test/ruby/test_time.rb
@@ -3,6 +3,7 @@
 require 'delegate'
 require 'timeout'
 require 'delegate'
+require 'rbconfig/sizeof'
 
 class TestTime < Test::Unit::TestCase
   def setup
@@ -1088,6 +1089,12 @@ def test_strftime_padding
     assert_equal("  -0:55:00", t.strftime("%_10::z"), bug6323)
     assert_equal("-000:55:00", t.strftime("%10::z"), bug6323)
     assert_equal("-000000:55", t.strftime("%10:::z"))
+
+    t = Time.at(0, in: 0)
+    max = RbConfig::LIMITS["INT_MAX"]
+    all_assertions_foreach(nil, *%W[%#{max}z %#{max}:z %#{max}::z %#{max}:::z]) do |z|
+      assert_raise(Errno::ERANGE) {t.strftime(z)}
+    end
   end
 
   def test_strftime_invalid_modifier
-- 
2.50.1 (Apple Git-155)

